[redis-streams support]

nicktylah opened the pull request:
- Adds methods to support [redis-streams](http://antirez.com/news/114) functionality (should probably leave in PR until that spec is finalized)
    - `XAdd`
    - `XAddExt`
    - `XLen`
    - `XRange`
    - `XRangeN`
    - `XRevRange`
    - `XRevRangeN`
    - `XRead`
    - `XReadN`

vmihailenco commented on commands.go:
```diff
@@ -152,6 +152,13 @@ type Cmdable interface {
+	XRange(key string, start, stop string) *XMessageSliceCmd
+	XRangeN(key string, start, stop string, count int64) *XMessageSliceCmd
+	XRead(block int64, streams []XStreamCheckpoint) *XStreamSliceCmd
+	XReadN(block, count int64, streams []XStreamCheckpoint) *XStreamSliceCmd
```
We almost certainly want `key string` to be the first argument so we can retrieve for in Redis Cluster in order to shard data - see https://github.com/go-redis/redis/blob/master/cluster.go#L543-L545. Having key in `XStreamCheckpoint` will complicate that code a lot for little/no reason.

nicktylah replied:
Solid point -- strangely though, it seems like the `XREAD` command is intended to allow reads from multiple streams at once, e.g. 
```
> XREAD BLOCK 5000 STREAMS mystream otherstream $ 1506935385635.0
```
Will the cluster support in this library allow this kind of multiplexing?

vmihailenco replied:
It expects that if there are multiple keys then all keys must be on the same shard.

vmihailenco commented on commands.go:
```diff
@@ -152,6 +152,13 @@ type Cmdable interface {
 	SUnion(keys ...string) *StringSliceCmd
 	SUnionStore(destination string, keys ...string) *IntCmd
+	XAdd(key string, id string, els ...interface{}) *StringCmd
+	XAddMaxLen(key string, id string, maxLen int64, approximate bool, els ...interface{}) *StringCmd
```
I am almost sure that `XAdd` will be extended later, .e.g. with MaxAge option https://github.com/antirez/redis/issues/4450. Also `redis.XAddMaxLen("stream", "12345.0", 100, false, "foo", "bar")` is not very readable. Can we change this to:

```
redis.XAddExt("stream", "12345.1", &redis.XAdd{
    MaxLen: 100,
    Approximate: true,
    MaxAge: time.Hour, // can be added later
}, "foo", "bar")
```

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1256,139 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
 
+// X represents stream message response.
+type X struct {
+	Key  string
```
Let's rename `Key` to `Stream` everywhere?

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1256,139 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
 //------------------------------------------------------------------------------
 
+// X represents stream message response.
+type X struct {
```
X -> XMessage please

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1256,139 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
+// XMessage represents messages within the stream response
+type XMessage struct {
+	ID string
+	Values map[string]interface{}
```
It looks inconsistent that we add `els ...interface{}`, but retrieve `Values map[string]interface{}`. If order of the message fields does not matter I would prefer to have `values map[string]interface{}` everywhere.

vmihailenco commented on parser.go:
```diff
@@ -111,6 +111,108 @@ func stringStructMapParser(rd *proto.Reader, n int64) (interface{}, error) {
 }
 
+// Implements proto.MultiBulkParse
+func xSliceParser(rd *proto.Reader, n int64) (interface{}, error) {
```
xSliceParser -> xStreamSliceParser?

vmihailenco commented on parser.go:
```diff
@@ -111,6 +111,108 @@ func stringStructMapParser(rd *proto.Reader, n int64) (interface{}, error) {
+			return nil, err
+		}
+
+		value, err := rd.ReadStringReply()
```
Why not `map[string]string` if value is always string?

nicktylah replied:
It should be an `interface{}`, but I couldn't find a method that would return the response as such

vmihailenco commented:
Hi,

Thanks for your contribution - code looks good, but there are some questions about API.

Also did you consider using `float64` for message id? It looks like wer are expected to do some math on the id so I wonder what is the best way to represent them

```
A final important thing to note about XRANGE is that, given that we receive the IDs in the reply, and the immediately successive ID is trivially obtained just incrementing the sequence part of the ID, it is possible to use XRANGE to incrementally iterate the whole stream, receiving for every call the specified number of elements. After the *SCAN family of commands in Redis, that allowed iteration of Redis data structures *despite* the fact they were not designed for being iterated, I avoided to make the same error again.
```

vmihailenco commented on commands.go:
```diff
@@ -152,6 +152,13 @@ type Cmdable interface {
+	XLen(key string) *IntCmd
+	XRange(key string, start, stop string) *XMessageSliceCmd
+	XRangeN(key string, start, stop string, count int64) *XMessageSliceCmd
+	XRead(block int64, streams []XStreamCheckpoint) *XStreamSliceCmd
```
Can we change to `block time.Duration`? That will make API cleaner.

nicktylah replied:
👍
What are your thoughts about `BLOCK 0` (block forever until new message(s) appear in the stream)? Currently this will wait as long as the client's `ReadTimeout` allows, then give an `i/o timeout` error. 

Should `XRead` be immune to these since it's intended functionality?

vmihailenco replied:
Yeah, library already allows you to set read timeout https://github.com/go-redis/redis/blob/master/commands.go#L947. Zero timeout is also supported.

nicktylah commented:
Thanks for the comments, I've updated some stuff accordingly.
I also chose to represent message IDs as their own type: 
``` type XMessageID struct {
  ID int64
  SeqID int
}
```
I think this will be a useful way to interact with them. Updated the affected parsers/commands to work with them

nicois commented:
Sequence ID is also 64 bit.
And don't forget XREVRANGE

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1257,177 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
+// and a sequence ID. E.g. 1511833543298-0
+type XMessageID struct {
+	ID int64
+	SeqID int
```
ID -> Utime or UnixTime
SeqID -> Seq

nicktylah replied:
You're not forced to make this a timestamp; e.g. `XADD stream 1 foo bar` is a valid command. For that reason, I don't think it makes sense to force it to be a time type

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1257,177 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
+	a[2] = "maxlen"
+	a[3] = "~"
+	a[4] = opts.MaxLen
+	if opts.AutoAssignID {
```
This could be `id != nil` and AutoAssignID can be removed

nicktylah commented on commands.go:
```diff
@@ -1249,6 +1257,177 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
+type XMessageID struct {
+	ID int64
+	SeqID int
+	End bool
```
I was using it to indicate `$`, but I'll use its absence as indicative of the end instead

nicktylah commented:
@nicois I can't run an `XREVRANGE` command against my `redis-streams` branch `redis-server`. Can you link to an example?

nicois commented:
@nicktylah the streams branch has been merged into unstable so pulling it won't give you the latest code any more.
Look at https://github.com/antirez/redis/commit/9bb18e54380250f3fb931028952379c3ab2dec29

nicktylah commented:
Pinging this PR for a re-review. Mergeable?

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1259,211 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
+type XMessageID struct {
+	ID int64
+	Seq int64
+}
```
I think API can be simplified a lot if `XMessageID` is defined as:

```
type XMessageID string

func (XMessageID) UnixSeq() (unix int64, seq int64)
```

vmihailenco commented on commands.go:
```diff
@@ -1249,6 +1259,211 @@ func (c *cmdable) SUnionStore(destination string, keys ...string) *IntCmd {
+	args[3] = stop
+	args[4] = "count"
+	args[5] = count
+	cmd := NewXMessageSliceCmd(args...)
```
This (and some other commands) can be written as:

```
cmd := NewXMessageSliceCmd(xrange, key, start, stop, "count", count)
```

vmihailenco commented:
I would like to make some changes, but what really concerns me is that there are no docs/beta releases at redis.io. Do you use unstable Redis Server? Can we wait until Redis team at least publishes beta release?

kassiansun commented:
Ouch, never seen this before... Should we merge the two MR, or do something based on either?

nicktylah commented:
Mmm good question. There's enough difference between the two PRs to extract some usefulness from both I think

skyrocknroll commented:
The docs are here https://redis.io/topics/streams-intro and RC is released now

vmihailenco commented:
Replaced by https://github.com/go-redis/redis/pull/795
